### PR TITLE
TPCQC: Adds map of TH1 in Tracks Task and Updates Utility

### PR DESCRIPTION
**Pull request by sbhawani**

### CODE REVIEW

File: Modules/TPC/run/tpcQCTracks_direct.json [hunk 1]
```diff
@@ -35,7 +33,7 @@
           "query": "inputTracks:TPC/TRACKS/0"
         },
         "taskParameters": {
-          "myOwnKey": "myOwnValue"
+          "EtaCut": "1.",  "nClusterCut": "60", "dEdxTot": "20."
```
Comment (wiechula) [inline Modules/TPC/run/tpcQCTracks_direct.json:36]: Perhaps a bit more verbose and consistent (either all with `Cut` or without)
```suggestion
          "AbsEtaCut": "1.",  "minNClusterCut": "60", "mindEdxTotCut": "20."
```
or
```suggestion
          "cutAbsEta": "1.",  "cutMinNCluster": "60", "cutMindEdxTott": "20."
```
or
```suggestion
          "AbsEta": "1.",  "minNCluster": "60", "mindEdxTot": "20."
```

File: Modules/TPC/src/Tracks.cxx [hunk 2]
```diff
@@ -68,14 +68,14 @@ void Tracks::monitorData(o2::framework::ProcessingContext& ctx)
 {
   using TrackType = std::vector<o2::tpc::TrackTPC>;
   auto tracks = ctx.inputs().get<TrackType>("inputTracks");
-  //using TracksType = gsl::span<o2::tpc::TrackTPC>;
-  //const auto tracks = ctx.inputs().get<TracksType>("inputTracks");
-  //ILOG(Info, Support) << "monitorData: " << tracks.size() << ENDM;
+  // using TracksType = gsl::span<o2::tpc::TrackTPC>;
+  // const auto tracks = ctx.inputs().get<TracksType>("inputTracks");
+  // ILOG(Info, Support) << "monitorData: " << tracks.size() << ENDM;
```
Comment (wiechula) [inline Modules/TPC/src/Tracks.cxx:73]: delete?

File: Modules/TPC/src/Tracks.cxx [hunk 3]
```diff
@@ -68,14 +68,14 @@ void Tracks::monitorData(o2::framework::ProcessingContext& ctx)
...
 
   for (auto const& track : tracks) {
     mQCTracks.processTrack(track);
   }
-  //mQCTracks.processAllTracks(tracks);
+  // mQCTracks.processAllTracks(tracks);
```
Comment (wiechula) [inline Modules/TPC/src/Tracks.cxx:78]: delete?

File: Modules/TPC/src/Tracks.cxx [hunk 4]
```diff
@@ -37,19 +38,18 @@ void Tracks::initialize(o2::framework::InitContext& /*ctx*/)
 {
   ILOG(Info, Support) << "initialize TPC Tracks QC task" << ENDM;
 
-  mQCTracks.initializeHistograms();
-  o2::tpc::qc::helpers::setStyleHistogram2D(mQCTracks.getHistograms2D());
+  float dEdxTot = getFromConfig<float>(mCustomParameters, "dEdxTot");
+  float EtaCutVal = getFromConfig<float>(mCustomParameters, "EtaCut");
+  int nClustCutVal = getFromConfig<int>(mCustomParameters, "nClusterCut");
```
Comment (wiechula) [inline Modules/TPC/src/Tracks.cxx:43]: See comment in json file, and
```suggestion
  const float dEdxTot = getFromConfig<float>(mCustomParameters, "dEdxTot");
  const float EtaCutVal = getFromConfig<float>(mCustomParameters, "EtaCut");
  const int nClustCutVal = getFromConfig<int>(mCustomParameters, "nClusterCut");
```

File: Modules/TPC/src/Utility.cxx [hunk 5]
```diff
@@ -150,4 +150,68 @@ std::unique_ptr<o2::tpc::internal::getWorkflowTPCInput_ret> clusterHandler(o2::f
...
 }
 
+template <typename T>
+T getFromConfig(const std::unordered_map<std::string, std::string>& params, const std::string_view name)
+{
+  const auto last = params.end();
+  const auto itParam = params.find(name.data());
+  T retVal;
+
+  std::string type = typeid(T).name();
```
Comment (wiechula) [inline Modules/TPC/src/Utility.cxx:160]: I think this is not uesd
```suggestion
```

File: Modules/TPC/src/Utility.cxx [hunk 6]
```diff
@@ -150,4 +150,68 @@ std::unique_ptr<o2::tpc::internal::getWorkflowTPCInput_ret> clusterHandler(o2::f
...
+  LOGP(info, "return value type: {}; return value: {}", typeid(retVal).name(), retVal);
+  return retVal;
+}
+
+std::string getFromConfig(const std::unordered_map<std::string, std::string>& params, const std::string_view name)
+{
+  const auto last = params.end();
+  const auto itParam = params.find(name.data());
+
+  std::string retVal;
```
Comment (wiechula) [inline Modules/TPC/src/Utility.cxx:196]: Initialize
```suggestion
  std::string retVal{};
```

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Modules/TPC/src/Tracks.cxx (hunk 3 above)
Comment (sbhawani) [reply Modules/TPC/src/Tracks.cxx]: done

File: Modules/TPC/src/Tracks.cxx (hunk 2 above)
Comment (sbhawani) [reply Modules/TPC/src/Tracks.cxx]: done

File: Modules/TPC/src/Utility.cxx [hunk 7]
```diff
@@ -150,4 +150,68 @@ std::unique_ptr<o2::tpc::internal::getWorkflowTPCInput_ret> clusterHandler(o2::f
...
+      }
+    } else {
+      LOG(error) << "Template type not supported";
+    }
+  }
+  LOGP(info, "return value type: {}; return value: {}", typeid(retVal).name(), retVal);
+  return retVal;
+}
+
+std::string getFromConfig(const std::unordered_map<std::string, std::string>& params, const std::string_view name)
```
[the remark this answers is not in the document]
Comment (sbhawani) [reply Modules/TPC/src/Utility.cxx]: Earlier for some unclear reasons it was not working but now it corrected and for strings as well.. implemented

File: Modules/TPC/src/Utility.cxx [hunk 8]
```diff
@@ -150,4 +150,68 @@ std::unique_ptr<o2::tpc::internal::getWorkflowTPCInput_ret> clusterHandler(o2::f
...
+    auto param = itParam->second;
+    LOG(info) << "param type: " << typeid(param).name() << " param: " << param << " param name: " << itParam->first;
+    std::string type = typeid(T).name();
+    if (type == "i") {
+      retVal = std::stoi(param);
+    } else if (type == "f") {
+      retVal = std::stof(param);
+    } else if (type == "d") {
+      retVal = std::stod(param);
+    } else if (type == "b") {
```
[the remark this answers is not in the document]
Comment (sbhawani) [reply Modules/TPC/src/Utility.cxx]: done

File: Modules/TPC/src/Utility.cxx (hunk 5 above)
Comment (sbhawani) [reply Modules/TPC/src/Utility.cxx]: yes i removed

File: Modules/TPC/src/Utility.cxx (hunk 6 above)
Comment (sbhawani) [reply Modules/TPC/src/Utility.cxx]: done

File: Modules/TPC/src/Utility.cxx [hunk 9]
```diff
@@ -150,4 +150,68 @@ std::unique_ptr<o2::tpc::internal::getWorkflowTPCInput_ret> clusterHandler(o2::f
   return std::move(retVal);
 }
 
+template <typename T>
+T getFromConfig(const std::unordered_map<std::string, std::string>& params, const std::string_view name)
+{
+  const auto last = params.end();
+  const auto itParam = params.find(name.data());
+  T retVal;
```
[the remark this answers is not in the document]
Comment (sbhawani) [reply Modules/TPC/src/Utility.cxx]: done

File: Modules/TPC/src/Utility.cxx [hunk 10]
```diff
@@ -150,4 +150,68 @@ std::unique_ptr<o2::tpc::internal::getWorkflowTPCInput_ret> clusterHandler(o2::f
...
+
+  std::string type = typeid(T).name();
+
+  if (itParam == last) {
+    LOGP(warning, "missing parameter {}", name.data());
+    LOGP(warning, "Please add '{}': '<value>' to the 'taskParameters'.", name.data());
+  } else {
+    auto param = itParam->second;
```
[the remark this answers is not in the document]
Comment (sbhawani) [reply Modules/TPC/src/Utility.cxx]: yes const has been ammended

File: Modules/TPC/run/tpcQCTracks_direct.json (hunk 1 above)
Comment (sbhawani) [reply Modules/TPC/run/tpcQCTracks_direct.json]: yeah newer version is there in the commit with  cut as prefix

File: Modules/TPC/src/Tracks.cxx (hunk 4 above)
Comment (sbhawani) [reply Modules/TPC/src/Tracks.cxx]: implemented

**Comment by sbhawani** [comment]
Hi @wiechula this PR can be merged now.. all the above comments have been implemented in the codes. thanks for comments

**Comment by Barthelemy** [comment]
we need @wiechula to approve the PR
I can force merge if there is an urgent need of course.

**Review by wiechula — APPROVED** [review]

**Pull request merged**